src/github.com/couchbase/sync_gateway/rest/handler.go

```diff
@@ -156,8 +170,29 @@ func (h *handler) invoke(method handlerMethod) error {
 		if err != nil {
 			return err
 		}
+
+		//If the Database is currently online process as normal
+		dbState := atomic.LoadUint32(&dbContext.State)
+		if (dbState == db.DBOnline) {
+			//Take a readlock for any method that does not run offline, taking a DB offline will
+			//block until all normal calls have returned and released their read locks.
+			if (!h.runOffline) {
+				//get a read lock on the dbContext
+				dbContext.AccessLock.RLock()
+
+				//defer releasing the dbContext until after the handler method returns
+				defer dbContext.AccessLock.RUnlock()
+			}
+		} else if (dbState == db.DBOffline) { //DB is offline
+			if (!h.runOffline) { //only handlers with runOffline true can run if DB is in Offline state
+				return base.HTTPErrorf(http.StatusServiceUnavailable, "DB is currently under maintenance")
+			}
+		} else { //DB is in transition state, no calls will be accepted until it is Online or Offline state
+			return base.HTTPErrorf(http.StatusServiceUnavailable, "DB is currently in neither online or offline state, try again later")
```

Comment:
This is also designed to stop e.g. _resync if the DB is not Offline.

Maybe "Database's state is transitioning - try again later"?